### PR TITLE
bug 1840616: kube-scheduler custom policy: do not use GeneralPredicates alongside its representants

### PR DESCRIPTION
**Pull request by bergerhoffer**
…its representants

Replaces #22521, and targeted against master to get to 4.5 as well.

Preview (internal): http://file.rdu.redhat.com/~ahoffer/2020/predicates-update/nodes/scheduling/nodes-scheduler-default.html#nodes-scheduler-default-about-understanding_nodes-scheduler-default

### CODE REVIEW

**Comment by bergerhoffer** [comment]
@kasturinarra Could you please review this update? Thanks!

Preview (internal): http://file.rdu.redhat.com/~ahoffer/2020/predicates-update/nodes/scheduling/nodes-scheduler-default.html#nodes-scheduler-default-about-understanding_nodes-scheduler-default

File: modules/nodes-scheduler-default-creating.adoc [hunk 1]
```diff
@@ -140,4 +137,3 @@ $ oc logs openshift-kube-scheduler-ip-10-0-141-29.ec2.internal | grep predicates
 
 Creating scheduler with fit predicates 'map[MaxGCEPDVolumeCount:{} MaxAzureDiskVolumeCount:{} CheckNodeUnschedulable:{} NoDiskConflict:{} NoVolumeZoneConflict:{} MatchNodeSelector:{} GeneralPredicates:{} MaxCSIVolumeCountPred:{} CheckVolumeBinding:{} MaxEBSVolumeCount:{} PodFitsResources:{} MatchInterPodAffinity:{} HostName:{} PodToleratesNodeTaints:{}]' and priority functions 'map[InterPodAffinityPriority:{} LeastRequestedPriority:{} ServiceSpreadingPriority:{} ImageLocalityPriority:{} SelectorSpreadPriority:{} EqualPriority:{} BalancedResourceAllocation:{} NodePreferAvoidPodsPriority:{} NodeAffinityPriority:{} TaintTolerationPriority:{}]'
```
Comment (kasturinarra) [inline modules/nodes-scheduler-default-creating.adoc:138]: @bergerhoffer one question here, when the above four predicates come under general predicates and not possible to declare it multiple times, would the log have reference to both general predicates and the four predicates which come under it  like shown aboave MatchNodeSelector , HostName  etc ? or would it have either of them ? Can you please confirm once ?

**Comment by kasturinarra** [comment]
I just looked at the doc which was pasted above and have few review comments.

1) under the section "creating a scheduler policy file" still see podFitsHostPorts, PodFitsResources, MatchNodeSelector, HostName being defined, can you please check and remove them ?
2) Under the point no.2 which reads "Edit the policy.cfg file to add and remove policies and predicates as needed." i still see all the four predicates present, could you please check this as well ?

Other than that i think the rest looks good to me, thanks !!

File: modules/nodes-scheduler-default-creating.adoc (hunk 1 above)
Comment (bergerhoffer) [reply modules/nodes-scheduler-default-creating.adoc:138]: @soltysh Can you confirm, should `PodFitsResources:{}`, `MatchNodeSelector:{}`, and `HostName:{}` be removed from this example output? Or would all 4 also be listed here?

**Comment by bergerhoffer** [comment]
@kasturinarra 

> under the section "creating a scheduler policy file" still see podFitsHostPorts, PodFitsResources, MatchNodeSelector, HostName being defined, can you please check and remove them ?

I believe that these are okay, because `GeneralPredicates` is not in that list. They are only not allowed when that is in the list as well.

> Under the point no.2 which reads "Edit the policy.cfg file to add and remove policies and predicates as needed." i still see all the four predicates present, could you please check this as well ?

I believe this is the same thing. Since  `GeneralPredicates` is not in this list, then it's okay for these 4 predicates to be listed.

Can you please confirm @soltysh?

**Comment by kasturinarra** [comment]
> @kasturinarra
> 
> > under the section "creating a scheduler policy file" still see podFitsHostPorts, PodFitsResources, MatchNodeSelector, HostName being defined, can you please check and remove them ?
> 
> I believe that these are okay, because `GeneralPredicates` is not in that list. They are only not allowed when that is in the list as well.
> 
> > Under the point no.2 which reads "Edit the policy.cfg file to add and remove policies and predicates as needed." i still see all the four predicates present, could you please check this as well ?
> 
> I believe this is the same thing. Since `GeneralPredicates` is not in this list, then it's okay for these 4 predicates to be listed.
> 
> Can you please confirm @soltysh?

@bergerhoffer i think it is better if we have consistency across the doc so that it does not confuse the user, however i will wait for soltysh to provide his inputs as well.

File: modules/nodes-scheduler-default-creating.adoc (hunk 1 above)
Comment (soltysh) [reply modules/nodes-scheduler-default-creating.adoc:138]: @kasturinarra is right, they should be removed as well.

Comment (soltysh) [reply modules/nodes-scheduler-default-creating.adoc:138]: Also while you're at it a few lines above here there's a typo:
`checks for the predoicates and priorites` 
should be 
`checks for the predicates and priorites`

**Comment by bergerhoffer** [comment]
@kasturinarra I've update the log message to remove those 3 predicates and fixed the typos.

As for the predicates showing up in step 1, as I mentioned, I think that's okay because GeneralPredicates isn't in the list. However, in looking closer, I'm not sure why the list of predicates and priorities in step 1's sample policy.cfg file is different than the output than the code block of the contents of the ConfigMap in step 2.

I think that we should either:

1. Make these two code blocks be in sync with their list of predicates and priorities, or
2. Just remove the last code block from step 2. It's an odd code block in that the above command doesn't actually output that or anything - you would need to enter another command to view the ConfigMap's contents like that.

What do you think? CC @soltysh

**Comment by bergerhoffer** [comment]
@kasturinarra @soltysh As I thought about it more, I went ahead and made some updates. I got rid of that second block, but updated the example policy.cfg file in step 1 to match the output in the log file in the last step. Let me know what you think!

Updated preview: https://predicates-update--ocpdocs.netlify.app/openshift-enterprise/latest/nodes/scheduling/nodes-scheduler-default.html#nodes-scheduler-default-creating_nodes-scheduler-default

**Comment by kasturinarra** [comment]
@bergerhoffer i looked at the preview document above and i see that the point where we edit the policy.cfg file still has PodFitsHostPorts, PodFitsResources, HostName. I think it is better to be consistent across the doc, so can we remove these three and add "General Predicates" here ?

Edit the policy.cfg file to add and remove policies and predicates as needed.

**Comment by bergerhoffer** [comment]
@kasturinarra Okay, so I updated so that the predicates/priorities in the "Edit the policy.cfg file to add and remove policies and predicates as needed." step now match the list used earlier in the procedure. They don't have any of those 4 predicates, only the `GeneralPredicates` now.

Let me know if this works.

File: modules/nodes-scheduler-default-creating.adoc [hunk 2]
```diff
@@ -14,7 +14,7 @@ You can control change the default scheduling behavior by creating a JSON file w
 
 To configure the scheduler policy:
 
-. Create the a JSON file named `policy.cfg` with the desired predicates and priorities. 
+. Create the a JSON file named `policy.cfg` with the desired predicates and priorities.
```
Comment (ingvagabund) [inline modules/nodes-scheduler-default-creating.adoc:17]: Create a JSON file ...

**Comment by ingvagabund** [comment]
@bergerhoffer @kasturinarra +1 for consistency

**Comment by ingvagabund** [comment]
/retitle bug 1840616: kube-scheduler custom policy: do not use GeneralPredicates alongside its representants

**Comment by kasturinarra** [comment]
@bergerhoffer looks good to me from the consistency point of view... Thanks !!

**Pull request merged**